### ISSUE TITLE
[UI Feature] Section for Katib Config in the UI

### ISSUE THREAD

**Issue by andreyvelich** [kubeflow/katib#1090]
/kind feature

Right now there is no section to edit/view `katib-config` in the UI.
What do you think will it be useful for Katib user to view and edit `katib-config` using UI ?

/cc @gaocegege @hougangliu @johnugeorge

**Comment by johnugeorge** [comment]
But this should be allowed only for admin user.

**Comment by kimwnasptd** [comment]
I've been thinking about how to implement this page while working on #1421 and also working on https://github.com/kubeflow/kubeflow/issues/5566.

As @johnugeorge mentioned the main caveat with this page is that is it shouldn't be just available to any user. From my understanding this is a two folded problem:
1. We need a mechanism to identify which user is using the web app
2. After we know which user is using the app [admin, normal user etc] then we need the establish the mechanism the UI will use to understand if it should show the page to that user or not

Once we know the user, then we can rely on K8s RBAC to check, via [SubjectAccessReviews](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/subject-access-review-v1/) if a user has permissions to view/edit the specific ConfigMap in the `kubeflow` namespace. In k8s we can use RBAC to give permission to a user only for specific [named objects](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources), and not for just all ConfigMaps in the namespace for example.

So the then the only tricky part is figuring out which user is using the app. In my mind this depends on whether the app is deployed in **standalone** or in **kubeflow** mode. I also tried to document the differences between these two modes in https://github.com/kubeflow/kubeflow/issues/5566.

### Kubeflow
Kubeflow expects that there will be a user identity, whether this comes from IAP, Dex etc. This identity is added to each request to the backend in a specific, well defined, header. So in this mode the app is aware of the logged in user. 

So in this mode the UI will simply rely on K8s RBAC to decide if it should allow the user to view the page and edit the ConfigMap.

### Standalone
The problem here is that there is no guarantee that there will be a logged in user. The web app might be exposed behind a login mechanism, like kubeflow expects or simply by using `kubectl port-forward`. So I'll expose my understanding based on the two cases:

#### Authn mechanim installed
We follow the same K8s RBAC approach we did for **kubeflow** mode. We will also expect that the authentication mechanism will be setting the user name in a specific header. The backend should be configurable to know which header to read this value from.

### No authn mechanism installed
In this case we expect that all the users are the same and have permission to do everything they want like CRUD Experiments, Trials etc.

**Comment by andreyvelich** [comment]
@kimwnasptd Does Kubeflow support admin user? As you mentioned before, we can give the admin user access to the `katib-config` ConfigMap in `kubeflow` namespace.

I think we should start with **No authn mechanism installed** in the Standalone app.

So Katib UI will have `ClusterRole` RBAC with access to all Experiments, Pods, ConfigMap and other necessary k8s resources.

If later we see requests from our users that they want to create their own auth mechanism with our Katib UI, we can extend the backend.

WDYT @kimwnasptd @johnugeorge @gaocegege ?

**Comment by gaocegege** [comment]
Yeah, I think maybe we can follow the design of k8s dashboard:

<img width="828" alt="截屏2021-02-01 下午3 28 39" src="https://user-images.githubusercontent.com/5100735/106427295-141f3080-64a2-11eb-8812-4947b7528c93.png">

**Comment by kimwnasptd** [comment]
I'll first try to expose my understanding and then I'll reiterate what I propose as next steps.

> @kimwnasptd Does Kubeflow support admin user?

Kubeflow does not expect the existence of specific users, like a user named `admin`. It only expects that
1. there will be an authentication mechanism in front of the Kubeflow cluster
2. this mechanism will be adding the name of the user to a header in each request that goes to Kubeflow and they underlying web apps

Then the web apps should use this header to know which user made the request and use `SubjectAccessReviews`, like [kubectl does](https://v1-16.docs.kubernetes.io/docs/reference/access-authn-authz/authorization/#checking-api-access), to ensure that this user is authorized to perform an action.

Kubeflow only defines a `kubeflow-admin`, `kubeflow-editor` and `kubeflow-viewer` ClusterRoles. Then it's up to the cluster admin to bind these cluster roles to specific users. This binding can be either for a specific namespace via `RoleBindings`, which is what `Profile` CRs do, or for the entire cluster with `ClusterRoleBindings`.

**Comment by kimwnasptd** [comment]
> As you mentioned before, we can give the admin user access to the katib-config ConfigMap in kubeflow namespace.

Based on the above explanation and understanding I think we should include this permission in the `kubeflow-admin` cluster role. Then it's up to the cluster admin to give this permission to individual users.

But this solution works only in case Kubeflow in installed, in order for these `ClusterRoles` to exist.

For the **standalone** mode, where we don't distinguish between who is using the app, we won't have to do anything extra. The backend will have the necessary permissions by default.

**Comment by kimwnasptd** [comment]
> I think we should start with No authn mechanism installed in the Standalone app.
> So Katib UI will have ClusterRole RBAC with access to all Experiments, Pods, ConfigMap and other necessary k8s resources.

Definitely! This will happen independently of whether the app is deployed as **standalone** or as part of **kubeflow**. The web app will always be the one that makes the requests to k8s so it should have such cluster-wide permissions.

The difference in the **kubeflow** mode is that the app/backend knows which user wants to make the request to k8s. If the backend would not take this information into consideration then this can lead to privilege escalation. For example a user could create an Experiment in a Namespace, via the backend, even if they don't have sufficient RBAC permissions.

So in **kubeflow** mode the backend will need to make a `SubjectAccessReview` first to ensure that the user that tries to make a request, via the backend, has the required K8s RBAC permissions.

**Comment by kimwnasptd** [comment]
So what I propose is:
* Always respect the K8s RBAC
* The app/backend must have always all the permissions for every namespace and CR [relative to AutoML WG]
* The UI should only allow a user to view the page for editing this `ConfigMap` only to those who have the k8s RBAC permissions
    * if we the UI is deployed with Kubeflow then it can figure out if a user has the required permissions, with `SubjectAccessReviews`
    * if the UI is deployed as **standalone** then it will let everyone edit this `ConfigMap`. In this case the only entity is the app which has by default the permission to view/edit this `ConfigMap`
* In the future, if people tell us that they want to use the **standalone** app but somehow be able to distinguish between users then we can discuss again how we could achieve this. This might require extending the backend as @andreyvelich and @gaocegege proposed. But again, lets not make any decision here until we see users request this

**Comment by andreyvelich** [comment]
Thank you for the explanation @kimwnasptd!
> Based on the above explanation and understanding I think we should include this permission in the kubeflow-admin cluster role. Then it's up to the cluster admin to give this permission to individual users.

Thus, even in the Kubeflow mode we can include Katib config to the frontend ? Since we can give the appropriate permission to the user.

> So in kubeflow mode the backend will need to make a SubjectAccessReview first to ensure that the user that tries to make a request, via the backend, has the required K8s RBAC permissions.

Sorry if I missed it before, but do we have any sign to the backend to notify that Katib UI is in Kubeflow mode ?

 
 
 > So what I propose is:
Sounds good, thank you @kimwnasptd!
